pkg/resource/v1/certconfig/desired.go

```diff
@@ -2,8 +2,375 @@ package certconfig
+			},
+		},
+		Spec: v1alpha1.CertConfigSpec{
+			Cert: v1alpha1.CertConfigSpecCert{
+				AllowBareDomains:    false,
+				ClusterComponent:    certName,
+				ClusterID:           clusterConfig.ClusterID,
+				CommonName:          clusterConfig.Domain.Prometheus,
+				DisableRegeneration: false,
+				TTL:                 clusterConfig.CertTTL,
+			},
+			VersionBundle: v1alpha1.CertConfigSpecVersionBundle{
+				Version: clusterConfig.VersionBundleVersion,
+			},
+		},
+	}
+}
+
+func newServiceAccountCertConfig(clusterConfig *cluster.Config, cert certs.Cert, projectName string) *v1alpha1.CertConfig {
+	certName := string(cert)
+	return &v1alpha1.CertConfig{
+		TypeMeta: apimetav1.TypeMeta{
+			Kind:       certKind,
+			APIVersion: certAPIVersion,
+		},
+		ObjectMeta: apimetav1.ObjectMeta{
+			Name: key.CertConfigName(clusterConfig.ClusterID, cert),
+			Labels: map[string]string{
+				label.Cluster:         clusterConfig.ClusterID,
+				label.LegacyClusterID: clusterConfig.ClusterID,
+				label.LegacyComponent: certName,
+				label.ManagedBy:       projectName,
+				label.Organization:    clusterConfig.Organization,
+			},
+		},
+		Spec: v1alpha1.CertConfigSpec{
+			Cert: v1alpha1.CertConfigSpecCert{
+				AllowBareDomains:    false,
+				ClusterComponent:    certName,
+				ClusterID:           clusterConfig.ClusterID,
+				CommonName:          clusterConfig.Domain.ServiceAccount,
+				DisableRegeneration: false,
+				TTL:                 clusterConfig.CertTTL,
+			},
+			VersionBundle: v1alpha1.CertConfigSpecVersionBundle{
+				Version: clusterConfig.VersionBundleVersion,
+			},
+		},
+	}
+}
+
+func newWorkerCertConfig(clusterConfig *cluster.Config, cert certs.Cert, projectName string) *v1alpha1.CertConfig {
+	certName := string(cert)
+	return &v1alpha1.CertConfig{
+		TypeMeta: apimetav1.TypeMeta{
+			Kind:       certKind,
+			APIVersion: certAPIVersion,
+		},
+		ObjectMeta: apimetav1.ObjectMeta{
+			Name: key.CertConfigName(clusterConfig.ClusterID, cert),
+			Labels: map[string]string{
+				label.Cluster:         clusterConfig.ClusterID,
+				label.LegacyClusterID: clusterConfig.ClusterID,
+				label.LegacyComponent: certName,
+				label.ManagedBy:       projectName,
+				label.Organization:    clusterConfig.Organization,
+			},
+		},
+		Spec: v1alpha1.CertConfigSpec{
+			Cert: v1alpha1.CertConfigSpecCert{
+				AllowBareDomains:    true,
+				AltNames:            kubeAltNames,
+				ClusterComponent:    certName,
+				ClusterID:           clusterConfig.ClusterID,
+				CommonName:          clusterConfig.Domain.Worker,
+				DisableRegeneration: false,
+				TTL:                 clusterConfig.CertTTL,
+			},
+			VersionBundle: v1alpha1.CertConfigSpecVersionBundle{
+				Version: clusterConfig.VersionBundleVersion,
+			},
+		},
+	}
+}
+
+func newServerDomain(commonDomain string, cert certs.Cert) (string, error) {
+	if !strings.Contains(commonDomain, ".") {
+		return "", microerror.Maskf(invalidConfigError, "commonDomain must be a valid domain")
+	}
```

Comment:
I believe we should validate custom objects somewhere else. I have an idea for a while about creating Validation Resource which would print an error about invalid CR and cancel reconciliation. We can sue `reconciliationcanceledcontext` for that. That would allow to not return error and declutter the code a little.

Reply:
Before having validations elsewhere, I'll leave this here.